pkg/processor/controlcommunication/controlcommunication.go

```diff
@@ -0,0 +1,195 @@
+func (c *ControlConsumer) BroadcastAndCloseSubscriptions(message *ControlMessage) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// do nothing if channels is an empty slice
+	if len(c.channels) == 0 {
+		return nil
+	}
+	// write message to the all channels
+	for _, channel := range c.channels {
+		go func(channel chan *ControlMessage, message *ControlMessage) {
+			defer func() {
+				// if the channel is closed before the message is read, it will result in a panic
+				if err := recover(); err != nil {
+					fmt.Println("Recovered in BroadcastAndCloseSubscriptions", err)
+				}
+			}()
+			channel <- message
+		}(channel, message)
+	}
+
+	// delete the channel from subscription
+	c.channels = make([]chan *ControlMessage, 0)
+
+	return nil
+}
+
+// Broadcast broadcasts a message to all subscribed channels
+func (c *ControlConsumer) Broadcast(message *ControlMessage) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(c.channels))
+	for _, channel := range c.channels {
+
+		go func(channel chan *ControlMessage, message *ControlMessage) {
+			defer func() {
+				// if the channel is closed before the message is read, it will result in a panic
+				if err := recover(); err != nil {
+					fmt.Println("Recovered in BroadcastAndCloseSubscriptions", err)
+				}
+			}()
+			defer wg.Done()
+			channel <- message
+		}(channel, message)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (c *ControlConsumer) addChannel(channel chan *ControlMessage) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.channels = append(c.channels, channel)
+}
+
+func (c *ControlConsumer) deleteChannel(channelToDelete chan *ControlMessage) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// remove the channel from the consumer
+	for i, channel := range c.channels {
+		if channel == channelToDelete {
+			c.channels = append(c.channels[:i], c.channels[i+1:]...)
+			break
+		}
+	}
+}
+
+type AbstractControlMessageBroker struct {
+	Consumers []*ControlConsumer
+}
+
+// NewAbstractControlMessageBroker creates a new abstract control message broker
+func NewAbstractControlMessageBroker() *AbstractControlMessageBroker {
+	// create a consumer for each control message kind
+	controlMessageKinds := getAllControlMessageKinds()
+	consumers := make([]*ControlConsumer, len(controlMessageKinds))
+	for index, kind := range controlMessageKinds {
+		consumers[index] = NewControlConsumer(kind)
+	}
+	return &AbstractControlMessageBroker{
+		Consumers: consumers,
+	}
+}
+
+func (acmb *AbstractControlMessageBroker) WriteControlMessage(message *ControlMessage) error {
+	return nil
+}
+
+func (acmb *AbstractControlMessageBroker) ReadControlMessage(reader *bufio.Reader) (*ControlMessage, error) {
+	return nil, nil
+}
+
+func (acmb *AbstractControlMessageBroker) SendToConsumers(message *ControlMessage) error {
+
+	for _, consumer := range acmb.Consumers {
+		if consumer.GetKind() == message.Kind {
+			switch message.Kind {
+			case DrainDoneMessageKind:
```

Comment:
Please add a comment describing why we need different handling here.
Think of a future developer seeing this and trying to understand.